### PR TITLE
chore(vitest): caching and add apperror checks

### PR DESCRIPTION
**Pull request by jurriaanr**
Adds a method of intermediate caching some passes to make reruns faster. Adds pass to check for AppError typing

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -26,6 +26,7 @@
 	},
 	"workspaces": ["packages/*", "apps/*"],
 	"volta": {
-		"node": "22.9.0"
+		"node": "22.11.0",
+		"typescript": "5.6.3"
```
Comment (dirkdev98) [inline package.json:30]: Why should this use another version than is installed locally in the project?

Comment (jurriaanr) [reply package.json]: It shouldn't its just what it is automatically pinned to when i pinned typescript

**Review by dirkdev98 — COMMENTED** [review]
```patch
Subject: [PATCH] chore(compas-convert): handle `t` usages, add non-nill assertions around not narrowable expressions
---
Index: packages/compas-convert/src/passes/convert-test-files.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/compas-convert/src/passes/convert-test-files.ts b/packages/compas-convert/src/passes/convert-test-files.ts
--- a/packages/compas-convert/src/passes/convert-test-files.ts	(revision 0d58aeed8eed9189e069692df17beb0a69d3eb2c)
+++ b/packages/compas-convert/src/passes/convert-test-files.ts	(date 1733473713633)
@@ -2,7 +2,6 @@
 import { Node } from "ts-morph";
 import { ts } from "ts-morph";
 import type { Context } from "../context.js";
-import { addPendingImport } from "../shared/imports.js";
 import SyntaxKind = ts.SyntaxKind;
 
 /**
@@ -49,14 +48,6 @@
 		)
 		?.remove();
 
-	// re-add import for newTestEvent
-	addPendingImport(context, sourceFile, "@compas/cli", "newTestEvent", false);
-	addPendingImport(context, sourceFile, "@compas/stdlib", "newLogger", false);
-	addPendingImport(context, sourceFile, "vitest", "expect", false);
-	addPendingImport(context, sourceFile, "vitest", "beforeAll", false);
-	addPendingImport(context, sourceFile, "vitest", "describe", false);
-	addPendingImport(context, sourceFile, "vitest", "test", false);
-
 	// go over each expression statement in file
 	for (const statement of sourceFile.getStatements()) {
 		if (statement.isKind(SyntaxKind.ExpressionStatement)) {
Index: packages/compas-convert/src/passes/assert-is-app-error-in-test-files.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/compas-convert/src/passes/assert-is-app-error-in-test-files.ts b/packages/compas-convert/src/passes/assert-is-app-error-in-test-files.ts
--- a/packages/compas-convert/src/passes/assert-is-app-error-in-test-files.ts	(revision 0d58aeed8eed9189e069692df17beb0a69d3eb2c)
+++ b/packages/compas-convert/src/passes/assert-is-app-error-in-test-files.ts	(date 1733470861036)
@@ -21,7 +21,8 @@
 	while (
 		(nextDiagnostic = getNextObjectOfTypeUnknownDiagnostic(sourceFile, lastPosition))
 	) {
-		lastPosition = nextDiagnostic.getStart()!;
+		// Move past the full error
+		lastPosition = nextDiagnostic.getStart()! + nextDiagnostic.getLength()! + 1;
 
 		// For some reason, we can't find the position of the diagnostic.
 		const expression = sourceFile.getDescendantAtPos(nextDiagnostic.getStart()!);
@@ -45,8 +46,7 @@
 		}
 
 		// The diagnostic text is not always helpful, so retrieve the expression from the file contents.
-		// -  "'user.passwordLogin' is possible 'undefined'"
-		// -  "Object is possible 'undefined'"
+		// -  "'e' is of type unknown'"
 		const expressionMatch = sourceFile
 			.getFullText()
 			.slice(
@@ -55,18 +55,21 @@
 			);
 
 		try {
-			const keyOrInfoRegex = new RegExp(`${expressionMatch}\\.(?:key|info)\\b`);
+			const keyOrInfoRegex = new RegExp(`${expressionMatch}\\.(?:key|info|status)\\b`);
 
 			if (
 				parentStatement.getText().match(keyOrInfoRegex) &&
 				isInCatchClause(parentStatement)
 			) {
+				const functionCallToInsert = `assertIsAppError(${expressionMatch});\n\n`;
+
 				// Assert is AppError does a type-narrowing assertion; meaning that it guarantees
-				// typescript,3 that it would throw and thus prevents execution of the normal code-path.
-				sourceFile.insertText(
-					parentStatement.getStart(true),
-					`assertIsAppError(${expressionMatch});\n\n`,
-				);
+				// TypeScript, that it would throw and thus prevents execution of the normal code-path.
+				sourceFile.insertText(parentStatement.getStart(true), functionCallToInsert);
+
+				// Make sure to the offset of the inserted call as well, so we don't stay in an infinite
+				// loop
+				lastPosition += functionCallToInsert.length;
 			}
 			// eslint-disable-next-line unused-imports/no-unused-vars
 		} catch (e) {
Index: packages/compas-convert/src/shared/cache.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/compas-convert/src/shared/cache.ts b/packages/compas-convert/src/shared/cache.ts
--- a/packages/compas-convert/src/shared/cache.ts	(revision 0d58aeed8eed9189e069692df17beb0a69d3eb2c)
+++ b/packages/compas-convert/src/shared/cache.ts	(date 1733472751528)
@@ -14,7 +14,7 @@
 		const fileDir = path
 			.dirname(this.sourceFile.getFilePath())
 			.replace(this.context.outputDirectory, "");
-		const hash = createHash("md5").update(this.sourceFile.getText()).digest("hex");
+		const hash = createHash("md5").update(this.sourceFile.getFullText()).digest("hex");
 		this.cacheFile = `${this.context.cacheDirectory}${fileDir}/${hash}`;
 	}
 
@@ -23,12 +23,13 @@
 			return false;
 		}
 
-		this.sourceFile.replaceWithText(readFileSync(this.cacheFile).toString("utf-8"));
+		this.sourceFile.replaceWithText(readFileSync(this.cacheFile, "utf-8"));
+		this.sourceFile.saveSync();
 		return true;
 	}
 
 	store() {
 		mkdirSync(path.dirname(this.cacheFile), { recursive: true });
-		writeFileSync(this.cacheFile, this.sourceFile.getText());
+		writeFileSync(this.cacheFile, this.sourceFile.getFullText());
 	}
 }
Index: packages/compas-convert/src/passes/add-common-imports.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/compas-convert/src/passes/add-common-imports.ts b/packages/compas-convert/src/passes/add-common-imports.ts
--- a/packages/compas-convert/src/passes/add-common-imports.ts	(revision 0d58aeed8eed9189e069692df17beb0a69d3eb2c)
+++ b/packages/compas-convert/src/passes/add-common-imports.ts	(date 1733473035254)
@@ -29,6 +29,13 @@
 		CONVERT_UTIL.assertNotNil,
 		false,
 	);
+	addPendingImport(
+		context,
+		sourceFile,
+		resolveRelativeImport(context, sourceFile, CONVERT_UTIL_PATH),
+		CONVERT_UTIL.assertIsAppError,
+		false,
+	);
 
 	addPendingImport(context, sourceFile, "@compas/stdlib", "InsightEvent", true);
 	addPendingImport(context, sourceFile, "@compas/stdlib", "Logger", true);
@@ -63,4 +70,14 @@
 	addPendingImport(context, sourceFile, "axios", "AxiosInstance", true);
 	addPendingImport(context, sourceFile, "axios", "AxiosRequestConfig", true);
 	addPendingImport(context, sourceFile, "axios", "AxiosError", true);
+
+	if (sourceFile.getFilePath().endsWith(".test.ts")) {
+		// re-add import for newTestEvent
+		addPendingImport(context, sourceFile, "@compas/cli", "newTestEvent", false);
+		addPendingImport(context, sourceFile, "@compas/stdlib", "newLogger", false);
+		addPendingImport(context, sourceFile, "vitest", "expect", false);
+		addPendingImport(context, sourceFile, "vitest", "beforeAll", false);
+		addPendingImport(context, sourceFile, "vitest", "describe", false);
+		addPendingImport(context, sourceFile, "vitest", "test", false);
+	}
 }
Index: packages/compas-convert/src/index.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/compas-convert/src/index.ts b/packages/compas-convert/src/index.ts
--- a/packages/compas-convert/src/index.ts	(revision 0d58aeed8eed9189e069692df17beb0a69d3eb2c)
+++ b/packages/compas-convert/src/index.ts	(date 1733473084342)
@@ -90,11 +90,10 @@
 ];
 
 const cachablePasses = [
-	"addCommonImports",
 	"convertTestFiles",
-	"convertTestConfig",
 	"notNilChecksInTestFiles",
 	"isAppErrorInTestFiles",
+	"finalizePendingImports",
 ];
 
 consola.start(`Converting ${path.relative(process.cwd(), resolvedInputDirectory)}`);
@@ -117,7 +116,7 @@
 
 			try {
 				const cache = new Cache(context, sourceFile);
-				if (cache.useIfExists() && cachablePasses.includes(pass.name)) {
+				if (cachablePasses.includes(pass.name) && cache.useIfExists()) {
 					continue;
 				}
 
```

Looks good, needed the above patch to get things going on different projects.

**Comment by jurriaanr** [comment]
Applied the patch manually (how to do it from here?)

**Comment by dirkdev98** [comment]
>  Applied the patch manually (how to do it from here?)

👍  
I don't think its possible. I was too time-crunchy to do individual suggestions :S

**Pull request merged**